IdentityCore/src/cache/mac/MSIDMacTokenCache.m

```diff
@@ -22,39 +22,448 @@
+    return result;
+}
+
+- (BOOL)saveWipeInfoWithContext:(id<MSIDRequestContext>)context
+                          error:(NSError **)error
+{
+    return NO;
+}
+
+- (NSDictionary *)wipeInfo:(id<MSIDRequestContext>)context
+                     error:(NSError **)error
 {
     return nil;
 }
 
+#pragma mark - Private
+
+- (void)addToItems:(nonnull NSMutableArray *)items
+    fromDictionary:(nonnull NSDictionary *)dictionary
+               key:(nonnull MSIDTokenCacheKey *)key
+{
+    MSIDToken *item = [dictionary objectForKey:[self keyWithoutAccount:key]];
+    if (item)
+    {
+        item = [item copy];
+        [items addObject:item];
+    }
+}
+
+- (void)addToItems:(nonnull NSMutableArray *)items
+         forUserId:(nonnull NSString *)userId
+            tokens:(nonnull NSDictionary *)tokens
+               key:(MSIDTokenCacheKey *)key
+{
+    NSDictionary *userTokens = [tokens objectForKey:userId];
+    if (!userTokens)
+    {
+        return;
+    }
+    
+    // Add items matching the key for this user
+    if (key.service)
+    {
+        [self addToItems:items fromDictionary:userTokens key:key];
+    }
+    else
+    {
+        for (id adkey in userTokens)
+        {
+            [self addToItems:items fromDictionary:userTokens key:adkey];
+        }
+    }
+}
+
+- (BOOL)validateCache:(NSDictionary *)dict
+                error:(NSError **)error
+{
+    RETURN_ERROR_IF_CONDITION_FALSE([dict isKindOfClass:[NSDictionary class]], MSIDErrorCacheBadFormat, @"Root level object of cache is not a NSDictionary!");
+    RETURN_ERROR_IF_CONDITION_FALSE(dict[@"version"], MSIDErrorCacheBadFormat, @"Missing version number from cache.");
+    RETURN_ERROR_IF_CONDITION_FALSE([dict[@"version"] floatValue] <= CURRENT_WRAPPER_CACHE_VERSION, MSIDErrorCacheBadFormat, @"Cache is a future unsupported version.");
+    
+    NSDictionary *cache = dict[@"tokenCache"];
+    RETURN_ERROR_IF_CONDITION_FALSE(cache, MSIDErrorCacheBadFormat, @"Missing token cache from data.");
+    RETURN_ERROR_IF_CONDITION_FALSE([cache isKindOfClass:[NSMutableDictionary class]], MSIDErrorCacheBadFormat, @"Cache is not a mutable dictionary!");
+    
+    NSDictionary *tokens = cache[@"tokens"];
+    
+    if (tokens)
+    {
+        RETURN_ERROR_IF_CONDITION_FALSE([tokens isKindOfClass:[NSMutableDictionary class]], MSIDErrorCacheBadFormat, @"tokens must be a mutable dictionary.");
+        for (id userId in tokens)
+        {
+            // On the second level we're expecting NSDictionaries keyed off of the user ids (an NSString*)
+            RETURN_ERROR_IF_CONDITION_FALSE([userId isKindOfClass:[NSString class]], MSIDErrorCacheBadFormat, @"User ID key is not of the expected class type.");
+            id userDict = [tokens objectForKey:userId];
+            RETURN_ERROR_IF_CONDITION_FALSE([userDict isKindOfClass:[NSMutableDictionary class]], MSIDErrorCacheBadFormat, @"User ID should have mutable dictionaries in the cache.");
+            
+            for (id key in userDict)
```

Comment:
can key be of some different type other MSIDTokenCacheKey?

Reply:
Yes, since we receive cache blob outside of our library. I added tests for validation logic.